src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc

```diff
@@ -293,6 +293,42 @@ BtreeLBAManager::init_cached_extent_ret BtreeLBAManager::init_cached_extent(
     });
 }
 
+BtreeLBAManager::scan_mappings_ret BtreeLBAManager::scan_mappings(
+  Transaction &t,
+  laddr_t begin,
+  laddr_t end,
+  scan_mappings_func_t &&f)
+{
+  return seastar::do_with(
```

Comment:
i think it's safe to put
```c++
  return get_root(t).safe_then(
    [=, &t, f=std::move(f)](LBANodeRef lbaroot) mutable {
      return lbaroot->scan_mappings(get_context(t), begin, end, f);
    });
```
here.

Reply:
Yes, but I find this to be clearer, and it's robust to a larger set of code transformations.

Reply:
okay. we can postpone the cleanup once the code is stablized a little bit if we can still apply this change by then.

Reply:
I can switch it now if you really prefer it.  I just don't find stashing the move into the lambda to be particularly clear.

Reply:
i understand. it's indeed a little bit weird. i think it's a new idiom we probably need to be used to when using seastar to postpone the cleanup of a non-copyable value.

Reply:
Heh, I went to swap it to your suggestion and it's actually wrong.  do_with will ensure that the captured values live until the contained future resolves, which is necessary here because scan_mappings will continue using f here long after the lambda that calls it has disappeared.

Reply:
> do_with will ensure that the captured values live until the contained future resolves,

i agree

> scan_mappings will continue using f here long after the lambda that calls it has disappeared.

agreed. i think it's `LBANode::scan_mappings()` 's signature should be 
```c++
  virtual scan_mappings_ret scan_mappings(
    op_context_t c,
    laddr_t begin,
    laddr_t end,
    scan_mappings_func_t &&f) = 0;
```
if we want to go with the simpler form. and it's the responsibility of, for instance, `LBAInternalNode::scan_mappings()` to capture-by-move `f`, and pass it down to the lambda expression in its `crimson::do_for_each()` call.

but, anyway, i think i suggest this change because i think it's has better readability and probably has better performance -- avoid an allocation of `unique_pointer<function>` on heap. both of them are kind of subjective.

Reply:
I like the way it works now and want to leave it alone.  scan_mappings is recursive and relatively complicated.  If it took f by rval, it would just use a do_with anyway at the top level.  Doing it this way enables the caller to do something clever or simply reuse f if it wants to.